man/man5/rlm_passwd.5
.\"     # DS - begin display
.de DS
.RS
.nf
.sp
..
.\"     # DE - end display
.de DE
.fi
.RE
.sp
..
.TH rlm_passwd 5 "20 January 2015" "" "FreeRADIUS Module"
.SH NAME
rlm_passwd \- FreeRADIUS Module
.SH DESCRIPTION
The \fIrlm_passwd\fP module provides authorization via files similar
in format to /etc/passwd.
.PP
This module allows you to retrieve any account information from any
files with passwd-like format (/etc/passwd, /etc/group,
smbpasswd, .htpasswd, etc). Every field of the file may be mapped to
a RADIUS attribute, with one of the fields used as a key.
.PP
The module reads the file when it initializes, and caches the data in
memory.  This makes it very fast, even for files with thousands of
lines. To re-read the file the module will need to be reloaded with
\fIradmin(8)\fP, or the server will need to be sent a SIGHUP, as
dynamic updates are not supported.
.PP
.SH CONFIGURATION
The configuration item(s):
.IP allow_multiple_keys
If set to 'yes', and more than one record in file matches the request,
then the attributes from all records will be used. If set to 'no' (the
default) the module will warn about duplicated records.
.IP delimiter\ =\ ":"
The character to use as a delimiter between fields.  The default is
":"
.IP filename
The path to the file.
.IP format
The format of the fields in the file, given as an example line from
the file, with the content of the fields as the RADIUS attributes
which the fields map to. The fields are separated by the ':' character
in the configuration (no matter what is configured for the 'delimiter'
option).
.IP hash_size
The size of the hash table. A larger value means less probability of a
collision so records will be found faster, at the expense of greater
memory usage. Having a hash_size in the range of 30-100% of the number
of passwd file records is reasonable.
.IP ignore_empty
When set to "yes", the default, empty fields in the input will be
skipped and the RADIUS attribute will not be added.  By setting this
value to "no", all attributes in the format list will always be added,
even if they have no value.
.IP ignore_nislike
If set to 'yes', then all records from the file beginning with the '+'
sign will be ignored.  The default is 'no'.
.PP
.SH FORMAT
The \fIformat\fP option controls how lines are read from the file, and
which fields are mapped to which RADIUS attributes.
.PP
The key field is the field being searched for within the file. It is
normally signified by being preceded with a '*' character, which
indicates that the field has only one key, like the /etc/passwd file.
The key field may instead be preceded with '*,', which indicates that
the field has multiple possible comma-separated keys, such as when
searching the /etc/group file.
.PP
The other fields signify RADIUS attributes. By default they will be
added as a control attribute list.
.PP
To add an attribute to the RADIUS request (as though it had been sent
by the NAS), prefix the attribute name in the "format" string with the
'~' character.
.PP
To add an attribute to the RADIUS reply (to be sent back to the NAS),
prefix the attribute name in the "format" string with the '='
character.
.PP
.SH EXAMPLES
.DS
format = "*User-Name:Crypt-Password:"
.DE
.IP
For a file the looks similar to /etc/passwd. The first field,
User-Name, is the key to look up in the file. When the record is
found, a control attribute, 'Crypt-Password', will be added with the
contents of the second field. (Note this will not work with shadow
passwords.)
.PP
.DS
format = "My-Group:::*,User-Name"
.DE
.IP
Parse a file similar to the /etc/group file.  An entry matches a
request when the name in a User-Name attribute exists in the
comma-separated list of a line in the file.  When an entry matches, 
a "My-Group" attribute will be created and added to the control
items for the request.  The value of that attribute will be taken from
the first field of the matching line in the file.
.IP
The ":::" in the format string means that there are extra two fields
in the line, in between the group name and list of user names.  Those
fields do not map to any RADIUS attribute, and are therefore ignored.
.IP
For this example to work in practice, you will have to add the
My-Group attribute to the dictionary file.  See \fIdictionary(5)\fP
for details on how this may be done.
.PP
.DS
format = "~My-Group:::*,User-Name"
.DE
.IP
Similar to the previous entry, except the My-Group attribute is added
to the request, as though it was sent by the NAS.
.PP
.SH SECTIONS
.BR authorize
.PP
.SH FILES
.I /etc/raddb/mods-available/passwd
.PP
.SH "SEE ALSO"
.BR radiusd (8),
.BR radiusd.conf (5),
.BR radmin (8),
.BR dictionary (5),
.BR rlm_unix (5)
.SH AUTHOR
Alan DeKok <aland@freeradius.org>, Matthew Newton
<matthew@newtoncomputing.co.uk>.



doc/ChangeLog
FreeRADIUS 2.2.9 Wednesday 30 Sep 2015 17:00:00 EDT, urgency=medium
	Feature improvements
	* None.

	Bug fixes
	* Fix Tunnel-Password crash from home server.
	  Found by Denis Andzakovic.
	* Fix timer issue when proxying.
	* Update EAP-TTLS so that MPPE keys are correctly
	  calculated with TLSv1.2.
	* Always delete MS-MPPE-* from the TTLS inner tunnel.
	  This allows TTLS / EAP-MSCHAPv2 to work.
	* Don't fall through in empty "case" statements.
	  Fixes #1274.

FreeRADIUS 2.2.8 Thursday 09 Jul 2015 11:00:00 EDT, urgency=medium
	Feature improvements
	* None.

	Bug fixes
	* Fixes for clients tied to virtual servers.  If there is
	  no "listen" section there, clients use the main "listen"
	  section.
	* Remove compiler warnings
	* Print out correct filenames in debug mode
	* Allow post-auth section to return "reject".  This turns
	  the response into Access-Reject.
	* Set X509_V_FLAG_CRL_CHECK_ALL, and add "check_all_crl"
	  to eap.conf.  Fixes oCert CVE-2015-4680.

FreeRADIUS 2.2.7 Wednesday 22 Apr 2015 14:00:00 EST, urgency=medium
	Feature improvements
	* Allow "eap" to be listed in Post-Auth-Type Reject
	  so that it sends EAP-Fail and Message-Authenticator.

	Bug fixes
	* Fix typo in code checking for blocked threads.
	  Closes #880.
	* Added more $(EXEEXT) to module utilities so that it
	  builds on Cygwin.  Closes #875.
	* Note that we don't need to generate ephemeral RSA keys.
	* Port detail file fixes from v3.
	* Use correct destination port for replies to DHCP relay.
	* rlm_perl can store multiple tagged attributes of the
	  same name.
	* Update EAP-TLS methods for TLSv1.2
	* Fix load-balance sections.  Closes #945

FreeRADIUS 2.2.6 Tuesday 18 Nov 2014 15:00:00 EST, urgency=medium
	Feature improvements
	* When supported by OpenSSL, allow TLS 1.1 and TLS 1.2
	  in EAP methods.

	Bug fixes
	* Fix redundant-load-balance blocks to try other modules in
	  the group if one fails.
	* Fix potential read into uninitialised memory in rlm_pap
	  when normalising octet type attributes containing password
	  hashes. This is very unlikely to happen in the wild.
	* Don't stop decoding DHCP options if we find a padding
	  option.
	* Define sig_t on systems which don't have it.  Closes #765
	* When clients are loaded from SQL, allow them to be tied
	  to a virtual server.
	* Prevent race conditions between fork and wait for child.
	  Patch from James Rouzier.
	* Allow UTF-8 characters in SQL.
	* Back-port udpfromto fixes from v3

FreeRADIUS 2.2.5 Monday 28 Apr 2014 15:20:00 EDT, urgency=medium
	Feature improvements
	* Update dictionary.terena.
	* expose server version via %v.  Patch from Alan Buxey.
	* Forbid running with vulnerable versions of OpenSSL.
	  See "allow_vulnerable_openssl" in the "security"
	  subsection of "radiusd.conf"
	* Catch underlying "heartbleed" problem, so that nothing bad
	  happens even when using a vulnerable version of OpenSSL.
	* Add dictionary.zte

	Bug fixes
	* Minor changes to build on Sun.
	* Print non-ASCII characters as octal in linelog.  Closes #578
	* close stdout in daemon mode.
	* Fix zombie period calculation.  Closes #579

FreeRADIUS 2.2.4 Wednesday 19 Mar 2014 13:20:00 EDT, urgency=medium
	Feature improvements
	* A "panic_action" can be set to have the server dump a gdb
	  log on SEGV or other fatal error.
	* allow radmin command "set module status <module> <code>"
	  which can be used to forcibly enable/disable modules.

	Bug fixes
	* If the server fails to bind() after fork(), that is now
	  reported to the parent, which exits with an error.
	* Session / delay times in MySQL are unsigned int.
	* Use --tag=CC for libtool.  Closes #497.  Because libtool
	  is too stupid to notice that compiling means compilation.
	* Fix bug when copying attributes for vendors > 32767
	* Fix behaviour on FreeBSD where sending packets from an interface
	  bound to an IP address would fail when the server was built with
	  udpfromto.
	* Don't fail config check if were listening on an IP which is
	  also a home server. Some deployments have valid reasons
	  to loop packets back to another virtual server.
	* Use correct port when DHCP relaying.
	* Set source IP address for DHCP packets from DHCP-Server-IP-Address,
	  or DHCP-DHCP-Server-Identifier, if we're unable to otherwise
	  determine the source IP.

FreeRADIUS 2.2.3 Wednesday 11 Dec 2013 15:00:00 EST, urgency=medium
	Feature improvements
	* Added dictionary.efficientip, dictionary.alcatel-lucent-aaa
	* Allow zero length DN strings in rlm_ldap.
	* If Password-With-Header has no header, assume it is
	  Cleartext-Password.

	Bug fixes
	* Make the server build when DHCP is enabled
	* Don't crash if there's no Post-Proxy-Type Reject.
	* Use correct fields for X509 attributes in certificates
	* Install threads.h making it possible to link against the
	  installed headers again.
	* Initialize SSL once in "main", instead of rlm_eap_tls.
	  Some client libraries may need SSL.

FreeRADIUS 2.2.2 Wednesday 30 Oct 2013 9:30:00 DST, urgency=medium
	Feature improvements
	* Add "timeout" to exec, and "ntlm_auth_timeout" to mschap.
	  So that run-away child processes are caught earlier.
	* Print out thread number for "unresponsive child".

	Bug fixes
	* Fix erroneous fall-through in "case" statements
	* Fix priority handling in new module handling code
	* Fix threading issue with Perl.  Closes #436
	* Fix EAP-TLS check_cert_issuer when X509v2 extensions
	  existed.  Patch from David Wood.
	* Fix pointer references in rlm_python.
	* Fix "unresponsive child" issue when proxying.
	* Set log output correctly when using -l.
	  Fix ported from 3.0.0.
	* Buffer debug output when threaded, so that text from
	  different threads isn't interspersed.
	* Fix SEGV in rlm_perl when using dynamic expansions.
	* Fix build for OSX Mavericks, which hid the header files
	  in a magical place.
	* Port DHCP fixes from 3.0.

FreeRADIUS 2.2.1 Tuesday 17 Sep 2013 12:00:00 CEST, urgency=medium
	Feature improvements

src/modules/rlm_passwd/rlm_passwd.c
/*
 *   This program is is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or (at
 *   your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 */

/**
 * $Id$
 * @file rlm_passwd.c
 * @brief Enables authentication against unix passwd files.
 *
 * @copyright 2000,2006  The FreeRADIUS server project
 */
RCSID("$Id$")

#include <freeradius-devel/radiusd.h>
#include <freeradius-devel/modules.h>
#include <freeradius-devel/rad_assert.h>

struct mypasswd {
	struct mypasswd *next;
	char *listflag;
	char *field[1];
};

struct hashtable {
	int tablesize;
	int keyfield;
	int nfields;
	int islist;
	int ignorenis;
	char * filename;
	struct mypasswd **table;
	char buffer[1024];
	FILE *fp;
	char delimiter;
};


#ifdef TEST

#define rad_malloc(s) malloc(s)

void printpw(struct mypasswd *pw, int nfields){
	int i;
	if (pw) {
		for( i = 0; i < nfields; i++ ) printf("%s:", pw->field[i]);
		printf("\n");
	}
	else printf ("Not found\n");
	fflush(stdout);
}
#endif


static struct mypasswd * mypasswd_malloc(char const* buffer, int nfields, size_t* len)
{
	struct mypasswd *t;
	/* reserve memory for (struct mypasswd) + listflag (nfields * sizeof (char*)) +
	** fields (nfields * sizeof (char)) + strlen (inst->format) + 1 */

	*len=sizeof (struct mypasswd) + nfields * sizeof (char*) + nfields * sizeof (char ) + strlen(buffer) + 1;
	t = (struct mypasswd *) rad_malloc(*len);
	if (t) memset(t, 0, *len);
	return (t);
}

static int string_to_entry(char const* string, int nfields, char delimiter,
			   struct mypasswd *passwd, size_t bufferlen)
{
	char *str;
	size_t len, i;
	int fn=0;
	char *data_beg;


	len = strlen(string);
	if(!len) return 0;
	if (string[len-1] == '\n') len--;
	if(!len) return 0;
	if (string[len-1] == '\r') len--;
	if(!len) return 0;
	if (!len || !passwd ||
	    bufferlen < (len + nfields * sizeof (char*) + nfields * sizeof (char) + sizeof (struct mypasswd) + 1) ) return 0;
	passwd->next = NULL;
	data_beg=(char *)passwd + sizeof(struct mypasswd);
	str = data_beg + nfields * sizeof (char) + nfields * sizeof (char*);
	memcpy (str, string, len);
	str[len] = 0;
	passwd->field[fn++] = str;
	passwd->listflag = data_beg + nfields * sizeof (char *);
	for(i=0; i < len; i++){
		if (str[i] == delimiter) {
			str[i] = 0;
			passwd->field[fn++] = str + i + 1;
			if (fn == nfields) break;
		}
	}
	for (; fn < nfields; fn++) passwd->field[fn] = NULL;
	return len + nfields * sizeof (char) + nfields * sizeof (char*) + sizeof (struct mypasswd) + 1;
}


static void destroy_password (struct mypasswd * pass)
{
	struct mypasswd *p;
	while ((p=pass)!=NULL) {
		pass = pass->next;
		free(p);
	}
}


static unsigned int hash(char const * username, unsigned int tablesize)
{
	int h=1;
	while (*username) {
		h = h * 7907 + *username++;
	}
	return h%tablesize;
}

static void release_hash_table(struct hashtable * ht){
	int i;

	if (!ht) return;
	for (i = 0; i < ht->tablesize; i++)
		if (ht->table[i])
			destroy_password(ht->table[i]);
	if (ht->table) {
		free(ht->table);
		ht->table = NULL;
	}
	if (ht->fp) {
		fclose(ht->fp);
		ht->fp = NULL;
	}
	ht->tablesize = 0;
}

static void release_ht(struct hashtable * ht){
	if (!ht) return;
	release_hash_table(ht);
	if (ht->filename) {
		free(ht->filename);
		ht->filename = NULL;
	}
	free(ht);
}

static struct hashtable * build_hash_table (char const * file, int nfields,
					    int keyfield, int islist, int tablesize, int ignorenis, char delimiter)
{
	struct hashtable* ht;
	size_t len;
	unsigned int h;
	struct mypasswd *hashentry, *hashentry1;
	char *list;
	char *nextlist=0;
	int i;
	char buffer[1024];

	ht = (struct hashtable *) rad_malloc(sizeof(struct hashtable));
	if(!ht) {
		return NULL;
	}
	memset(ht, 0, sizeof(struct hashtable));
	ht->filename = strdup(file);
	if(!ht->filename) {
		free(ht);
		return NULL;
	}
	ht->tablesize = tablesize;
	ht->nfields = nfields;
	ht->keyfield = keyfield;
	ht->islist = islist;
	ht->ignorenis = ignorenis;
	if (delimiter) ht->delimiter = delimiter;
	else ht->delimiter = ':';
	if(!tablesize) return ht;
	if(!(ht->fp = fopen(file,"r"))) {
		free(ht->filename);
		free(ht);
		return NULL;
	}

	/*
	 *	@todo: This code is SHIT.  It's badly formatted.  It's
	 *	hard to understand.  It re-implements tons of things
	 *	which are already in the server core.
	 */
	memset(ht->buffer, 0, 1024);
	ht->table = (struct mypasswd **) rad_malloc (tablesize * sizeof(struct mypasswd *));
	if (!ht->table) {
		/*
		 * Unable allocate memory for hash table
		 * Still work without it
		 */
		ht->tablesize = 0;
		return ht;
	}
	memset(ht->table, 0, tablesize * sizeof(struct mypasswd *));
	while (fgets(buffer, 1024, ht->fp)) {
		if(*buffer && *buffer!='\n' && (!ignorenis || (*buffer != '+' && *buffer != '-')) ){
			if(!(hashentry = mypasswd_malloc(buffer, nfields, &len))){
				release_hash_table(ht);
				return ht;
			}
			len = string_to_entry(buffer, nfields, ht->delimiter, hashentry, len);
			if(!hashentry->field[keyfield] || *hashentry->field[keyfield] == '\0') {
				free(hashentry);
				continue;
			}

			if (islist) {
				list = hashentry->field[keyfield];
				for (nextlist = list; *nextlist && *nextlist!=','; nextlist++);
				if (*nextlist) *nextlist++ = 0;
				else nextlist = 0;
			}
			h = hash(hashentry->field[keyfield], tablesize);
			hashentry->next = ht->table[h];
			ht->table[h] = hashentry;
			if (islist) {
				for(list=nextlist; nextlist; list = nextlist){
					for (nextlist = list; *nextlist && *nextlist!=','; nextlist++);
					if (*nextlist) *nextlist++ = 0;
					else nextlist = 0;
					if(!(hashentry1 = mypasswd_malloc("", nfields, &len))){
						release_hash_table(ht);
						return ht;
					}
					for (i=0; i<nfields; i++) hashentry1->field[i] = hashentry->field[i];
					hashentry1->field[keyfield] = list;
					h = hash(list, tablesize);
					hashentry1->next = ht->table[h];
					ht->table[h] = hashentry1;
				}
			}
		}
	}
	fclose(ht->fp);
	ht->fp = NULL;
	return ht;
#undef passwd
}

static struct mypasswd * get_next(char *name, struct hashtable *ht,
				  struct mypasswd **last_found)
{
	struct mypasswd * passwd;
	struct mypasswd * hashentry;
	char buffer[1024];
	char *list, *nextlist;

	if (ht->tablesize > 0) {
		/* get saved address of next item to check from buffer */
		hashentry = *last_found;
		for (; hashentry; hashentry = hashentry->next) {
			if (!strcmp(hashentry->field[ht->keyfield], name)) {
				/* save new address */
				*last_found = hashentry->next;
				return hashentry;
			}
		}
		return NULL;
	}
	/*	printf("try to find in file\n"); */
	if (!ht->fp) return NULL;

	passwd = (struct mypasswd *) ht->buffer;

	while (fgets(buffer, 1024,ht->fp)) {
		if(*buffer && *buffer!='\n' && string_to_entry(buffer, ht->nfields, ht->delimiter, passwd, sizeof(ht->buffer)-1) &&
		   (!ht->ignorenis || (*buffer !='-' && *buffer != '+') ) ){
			if(!ht->islist) {
				if(!strcmp(passwd->field[ht->keyfield], name))
					return passwd;
			}
			else {
				for (list = passwd->field[ht->keyfield], nextlist = list; nextlist; list = nextlist) {
					for(nextlist = list; *nextlist && *nextlist!=','; nextlist++);
					if(!*nextlist) {
						nextlist = 0;
					} else {
						*nextlist++ = 0;
					}
					if (!strcmp(list, name)) {
						return passwd;
					}
				}
			}

		}
	}
	fclose(ht->fp);
	ht->fp = NULL;
	return NULL;
}

static struct mypasswd * get_pw_nam(char * name, struct hashtable* ht,
				    struct mypasswd **last_found)
{
	int h;
	struct mypasswd * hashentry;

	if (!ht || !name || *name == '\0') return NULL;
	*last_found = NULL;
	if (ht->tablesize > 0) {
		h = hash (name, ht->tablesize);
		for (hashentry = ht->table[h]; hashentry; hashentry = hashentry->next) {
			if (!strcmp(hashentry->field[ht->keyfield], name)){
				/* save address of next item to check into buffer */
				*last_found=hashentry->next;
				return hashentry;
			}
		}

		return NULL;
	}
	if (ht->fp) {
		fclose(ht->fp);
		ht->fp = NULL;
	}
	if (!(ht->fp=fopen(ht->filename, "r"))) return NULL;
	return get_next(name, ht, last_found);
}

#ifdef TEST

#define MALLOC_CHECK_ 1

int main(void){
	struct hashtable *ht;
	char *buffer;
	struct mypasswd* pw, *last_found;
	int i;

	ht = build_hash_table("/etc/group", 4, 3, 1, 100, 0, ":");
	if(!ht) {
		printf("Hash table not built\n");
		return -1;
	}
	for (i = 0; i < ht->tablesize; i++) {
		if (ht->table[i]) {
			printf("%d:\n", i);
			for (pw = ht->table[i]; pw; pw = pw->next) {
				printpw(pw, 4);
			}
		}
	}

	while(fgets(buffer, 1024, stdin)){
		buffer[strlen(buffer)-1] = 0;
		pw = get_pw_nam(buffer, ht, &last_found);
		printpw(pw,4);
		while ((pw = get_next(buffer, ht, &last_found))) printpw(pw,4);
	}
	release_ht(ht);
}

#else  /* TEST */
typedef struct rlm_passwd_t {
	struct hashtable	*ht;
	struct mypasswd		*pwdfmt;
	char const		*filename;
	char const		*format;
	char const		*delimiter;
	bool			allow_multiple;
	bool			ignore_nislike;
	uint32_t		hash_size;
	uint32_t		nfields;
	uint32_t		keyfield;
	uint32_t		listable;
	DICT_ATTR const		*keyattr;
	bool			ignore_empty;
} rlm_passwd_t;

static const CONF_PARSER module_config[] = {
	{ FR_CONF_OFFSET("filename", PW_TYPE_FILE_INPUT | PW_TYPE_REQUIRED, rlm_passwd_t, filename) },
	{ FR_CONF_OFFSET("format", PW_TYPE_STRING | PW_TYPE_REQUIRED, rlm_passwd_t, format) },
	{ FR_CONF_OFFSET("delimiter", PW_TYPE_STRING, rlm_passwd_t, delimiter), .dflt = ":" },

	{ FR_CONF_OFFSET("ignore_nislike", PW_TYPE_BOOLEAN, rlm_passwd_t, ignore_nislike), .dflt = "yes" },

	{ FR_CONF_OFFSET("ignore_empty", PW_TYPE_BOOLEAN, rlm_passwd_t, ignore_empty), .dflt = "yes" },

	{ FR_CONF_OFFSET("allow_multiple_keys", PW_TYPE_BOOLEAN, rlm_passwd_t, allow_multiple), .dflt = "no" },

	{ FR_CONF_OFFSET("hash_size", PW_TYPE_INTEGER, rlm_passwd_t, hash_size), .dflt = "100" },
	CONF_PARSER_TERMINATOR
};

static int mod_instantiate(CONF_SECTION *conf, void *instance)
{
	int nfields=0, keyfield=-1, listable=0;
	char const *s;
	char *lf=NULL; /* destination list flags temporary */
	size_t len;
	int i;
	DICT_ATTR const * da;
	rlm_passwd_t *inst = instance;

	rad_assert(inst->filename && *inst->filename);
	rad_assert(inst->format && *inst->format);

	if (inst->hash_size == 0) {
		cf_log_err_cs(conf, "Invalid value '0' for hash_size");
		return -1;
	}

	lf = talloc_typed_strdup(inst, inst->format);
	if ( !lf) {
		ERROR("rlm_passwd: memory allocation failed for lf");
		return -1;
	}
	memset(lf, 0, strlen(inst->format));
	s = inst->format - 1;
	do {
		if(s == inst->format - 1 || *s == ':'){
			if(*(s+1) == '*'){
				keyfield = nfields;
				s++;
			}
			if(*(s+1) == ','){
				listable = 1;
				s++;
			}
			if(*(s+1) == '='){
				lf[nfields]=1;
				s++;
			}
			if(*(s+1) == '~'){
				lf[nfields]=2;
				s++;
			}
			nfields++;
		}
		s++;
	}while(*s);
	if(keyfield < 0) {
		cf_log_err_cs(conf, "no field marked as key in format: %s",
			      inst->format);
		return -1;
	}
	if (! (inst->ht = build_hash_table (inst->filename, nfields, keyfield, listable, inst->hash_size, inst->ignore_nislike, *inst->delimiter)) ){
		ERROR("rlm_passwd: can't build hashtable from passwd file");
		return -1;
	}
	if (! (inst->pwdfmt = mypasswd_malloc(inst->format, nfields, &len)) ){
		ERROR("rlm_passwd: memory allocation failed");
		release_ht(inst->ht);
		inst->ht = NULL;
		return -1;
	}
	if (!string_to_entry(inst->format, nfields, ':', inst->pwdfmt , len)) {
		ERROR("rlm_passwd: unable to convert format entry");
		release_ht(inst->ht);
		inst->ht = NULL;
		return -1;
	}

	memcpy(inst->pwdfmt->listflag, lf, nfields);

	talloc_free(lf);
	for (i=0; i<nfields; i++) {
		if (*inst->pwdfmt->field[i] == '*') inst->pwdfmt->field[i]++;
		if (*inst->pwdfmt->field[i] == ',') inst->pwdfmt->field[i]++;
		if (*inst->pwdfmt->field[i] == '=') inst->pwdfmt->field[i]++;
		if (*inst->pwdfmt->field[i] == '~') inst->pwdfmt->field[i]++;
	}
	if (!*inst->pwdfmt->field[keyfield]) {
		cf_log_err_cs(conf, "key field is empty");
		release_ht(inst->ht);
		inst->ht = NULL;
		return -1;
	}
	if (! (da = dict_attrbyname (inst->pwdfmt->field[keyfield])) ) {
		ERROR("rlm_passwd: unable to resolve attribute: %s", inst->pwdfmt->field[keyfield]);
		release_ht(inst->ht);
		inst->ht = NULL;
		return -1;
	}
	inst->keyattr = da;
	inst->nfields = nfields;
	inst->keyfield = keyfield;
	inst->listable = listable;
	DEBUG2("rlm_passwd: nfields: %d keyfield %d(%s) listable: %s", nfields, keyfield, inst->pwdfmt->field[keyfield], listable?"yes":"no");
	return 0;

#undef inst
}

static int mod_detach (void *instance) {
#define inst ((rlm_passwd_t *)instance)
	if(inst->ht) {
		release_ht(inst->ht);
		inst->ht = NULL;
	}
	free(inst->pwdfmt);
	return 0;
#undef inst
}

static void addresult (TALLOC_CTX *ctx, rlm_passwd_t *inst, REQUEST *request,
		       VALUE_PAIR **vps, struct mypasswd * pw, char when, char const *listname)
{
	uint32_t i;
	VALUE_PAIR *vp;

	for (i = 0; i < inst->nfields; i++) {
		if (inst->pwdfmt->field[i] && *inst->pwdfmt->field[i] && pw->field[i] && i != inst->keyfield  && inst->pwdfmt->listflag[i] == when) {
			if ( !inst->ignore_empty || pw->field[i][0] != 0 ) { /* if value in key/value pair is not empty */
				vp = fr_pair_make(ctx, vps, inst->pwdfmt->field[i], pw->field[i], T_OP_EQ);
				if (vp) {
					RDEBUG("Added %s: '%s' to %s ", inst->pwdfmt->field[i], pw->field[i], listname);
				}
			} else
				RDEBUG("NOOP %s: '%s' to %s ", inst->pwdfmt->field[i], pw->field[i], listname);
		}
	}
}

static rlm_rcode_t CC_HINT(nonnull) mod_passwd_map(void *instance, REQUEST *request)
{
#define inst ((rlm_passwd_t *)instance)
	char buffer[1024];
	VALUE_PAIR *key, *i;
	struct mypasswd * pw, *last_found;
	vp_cursor_t cursor;

	key = fr_pair_find_by_da(request->packet->vps, inst->keyattr, TAG_ANY);
	if (!key) {
		return RLM_MODULE_NOTFOUND;
	}

	for (i = fr_cursor_init(&cursor, &key);
	     i;
	     i = fr_cursor_next_by_num(&cursor, inst->keyattr->attr, inst->keyattr->vendor, TAG_ANY)) {
		/*
		 *	Ensure we have the string form of the attribute
		 */
		fr_pair_value_snprint(buffer, sizeof(buffer), i, 0);
		if (!(pw = get_pw_nam(buffer, inst->ht, &last_found)) ) {
			continue;
		}
		do {
			addresult(request, inst, request, &request->config, pw, 0, "config");
			addresult(request->reply, inst, request, &request->reply->vps, pw, 1, "reply_items");
			addresult(request->packet, inst, request, &request->packet->vps, pw, 2, "request_items");
		} while ((pw = get_next(buffer, inst->ht, &last_found)));

		if (!inst->allow_multiple) {
			break;
		}
	}

	return RLM_MODULE_OK;

#undef inst
}

extern module_t rlm_passwd;
module_t rlm_passwd = {
	.magic		= RLM_MODULE_INIT,
	.name		= "passwd",
	.type		= RLM_TYPE_HUP_SAFE,
	.inst_size	= sizeof(rlm_passwd_t),
	.config		= module_config,
	.instantiate	= mod_instantiate,
	.detach		= mod_detach,
	.methods = {
		[MOD_AUTHORIZE]		= mod_passwd_map,
		[MOD_ACCOUNTING]	= mod_passwd_map,
		[MOD_POST_AUTH]		= mod_passwd_map,
		[MOD_PRE_PROXY]		= mod_passwd_map,
		[MOD_POST_PROXY]  	= mod_passwd_map,
#ifdef WITH_COA
		[MOD_RECV_COA]		= mod_passwd_map,
		[MOD_SEND_COA]		= mod_passwd_map
#endif
	},
};
#endif /* TEST */
